### ISSUE TITLE
some chars are not displaying

### ISSUE THREAD

**Issue by mouadh-dev** [khaled-alshamaa/ar-php#29]
hi,
your project is awesome and it help me a lot, but there is a problem I didn't know if it's a bug or a mistake from me,
there are some chars that don't display in the output and the direction is always from left to right take a look

my code :
```
public function generatePoster(Request $request,Arabic $obj) {
        $title = $request->get('title');
        $title =$obj->utf8Glyphs($title,20,false,true);
        $imgPath = public_path('storage/ichecktemplate.jpg');
        $image = imagecreatefromjpeg($imgPath);
        $color = imagecolorallocate($image, 255, 255, 255);
        $fontSize = 30 ;
        $font = public_path('Almarai-Bold.ttf')
        $x = 640;
        $y = 364;
        imagettftext($image, $fontSize, 0, $x, $y, $color, $font, $title);
        imagejpeg($image, 'save.jpg', 100);
        }
```
my input: ابتثجحدحدذر زسشصض طعغفقك لمنهو يءأإآةؤ ئى
output:
![image](https://user-images.githubusercontent.com/62395675/172226027-82c6aeab-dfce-416c-9b69-0273c71117cd.png)

**Comment by muotaz** [comment]
السلام عليكم 

الرجاء الاطلاع على التكت التالية: 
https://github.com/khaled-alshamaa/ar-php/issues/25

**Comment by sakairyota** [comment]
Hi,

In our project, similer problem has occured.
And we found the partial condition of the problem.

We are using utf8Glyphs function.
```
$result = $arabic->utf8Glyphs("ٱد");
```
has produced invalid UTF-8 byte stream. 
```
unpack('C*', $result);
=> [
     1 => 217,
     2 => 32,
     3 => 239,
     4 => 186,
     5 => 169,
   ]
```
217, 32 is invalid UTF-8 sequence.

We found the following condition to reproduce the issue.
- The text is starting from non arabic char (characters not in range of u060C - u0652)
- When The char just before the first arabic char is an ascii char, it becomes white space ' '.
- When The char just before the first arabic char is a multi byte character, the squence becomes invalid (the last byte of the char is converted to white space byte)

For example "aد" becomes " ﺩ".
And "あد" becomes b"Òü ´║®" (invalid utf-8 squence)

Additionally ٱ (u0671) is arabic character but not in range u060C - u0652. "ٱد" (first example) also produce invalid squence .

Thank you.

**Comment by khaled-alshamaa** [comment]
Dear @mouadh-dev,

Please check this issue #25, the same problem has been reported before and we already fix it. Either download the latest development version from here in GitHub, or wait until we release the next version 6.3 within a week from now.

Thanks for your feedback 🤗

**Comment by mouadh-dev** [comment]
Dear @khaled-alshamaa,
thanks for your effort I appreciate that.

**Comment by khaled-alshamaa** [comment]
Dear @sakairyota,

Thanks again for bringing this issue to the table. It is just the right time when we are extensively reviewing that legacy `utf8Glyphs` part of the code. Well, your issue was not straightforward and touched on several parts. Therefore, I will divide and conquer ;-)

1. The `$cDec` in the `arIdentifier` method does not start from the 2nd byte because the initial `$i` value is -1
2. We have a method called `addGlyphs` to add any extra letter which does not exist yet in the library (just like the case of the letter ٱ that you mentioned). You can add it using the following line of code:
```php
$Arabic->addGlyphs('ٱ', 'FB50FB51FB50FB51', false, true);
```
> * The 2nd parameter is a string of 16 hexadecimal digits referring to the letter Unicode in the following order (ISOLATED FORM, FINAL FORM, INITIAL FORM, MEDIAL FORM)
> * The 3rd parameter tills that our added char can not connect to the following letter.
> * The 4th parameter tills that our added char can connect to the previous letter.

But when I tried that solution myself, I noticed another related issue. The `arIdentifier` does not recognize that letter as an Arabic letter! When I dug deep, I found that the range defined in that method was not complete and miss adding some extended Arabic characters. Therefore, I redefined it properly using [this reference](https://utf8-chartable.de/unicode-utf8-table.pl?start=1536&number=128&utf8=dec), and it is working fine now.

One final note, the new version will have an [external JSON file](https://github.com/khaled-alshamaa/ar-php/blob/master/src/data/ar_glyphs.json) where you can contribute easily by adding any Arabic characters not already included.

I will work now on the other issues you mentioned. Sorry to be late in handling these reported bugs. I am just waiting for my weekend to work on this side project.

**Comment by khaled-alshamaa** [comment]
Hi @sakairyota, here is me again! 

Thank you for carefully reviewing the utf8Glyphs method and for reporting this precise feedback with clear reproducing instructions. 

I am happy to let you know that we also [solved](https://github.com/khaled-alshamaa/ar-php/commit/19cd3934e092331fa0cb72bd80c9813a55959b3a) the issue you submitted when we have Arabic and non-Arabic chars in the same word (i.e., with no space separator) and the case of losing the first char when a string/line starts with a non-Arabic char.

**Comment by khaled-alshamaa** [comment]
Dear @sakairyota, please download the latest version directly from GitHub, test it, and close this issue if we resolved your reported problem properly.

**Comment by mouadh-dev** [comment]
Dear @khaled-alshamaa ,
This is my probleme and it's not resolved yet, still the direction rtl not working.

**Comment by khaled-alshamaa** [comment]
Dear @mouadh-dev,

I already resolved the problem you mentioned at the root of this thread in this [previous reply](https://github.com/khaled-alshamaa/ar-php/issues/29#issuecomment-1148934849).

Please note that you say nothing about any RTL problem, but let me guess from your screenshot that you are referring to the right align text issue, which is simply out of this library scope! 

It is related to the output function/library you are using. For example, the [imagettftext()](https://www.php.net/manual/en/function.imagettftext.php) has no parameter to control that. Well, there are some tricks to overcome that limitation by using the `imagettfbox()` functionality like the one described in [this stack overflow post](https://stackoverflow.com/questions/2444015/right-align-text-in-an-image-with-imagettftext-php).

I hope these notes give you enough clue to solve your problem ;-)

**Comment by mouadh-dev** [comment]
Dear @khaled-alshamaa ,
I am sorry if I didn't clarify my problem but I mention the direction in the issue,
and thanks for any help you gave to me.

**Comment by sakairyota** [comment]
Dear @khaled-alshamaa 

Thank you for response and the modification.

> The $cDec in the arIdentifier method does not start from the 2nd byte because the initial $i value is -1

Yes, It start with -1, but it is incremented in while condition and it becomes 0, and then `$ascii[$i + 1]` is evaluated. So I think the first `$cDec ` value is `$ascii[1]`.


New JSON format looks good with extensibility. But unfortunately all of our team members do not familiar with Arabic and Arabic scripts, so it is slightly difficult for us to use this (I learned Arabic script system a little in [Wikipedia](https://en.wikipedia.org/wiki/Arabic_script_in_Unicode) etc., but still it is difficult to find right form especially the letters in expanded block :( .


Solved version looks good. I'll try it.

> Sorry to be late in handling these reported bugs. 
I understand, and rather your quick response helps me. Thanks a lot.

**Comment by sakairyota** [comment]
@sakairyota 
I have tried 3 patterns, and 2 patterns are OK.  But 'ٱب' becomes 'ب&#x;' without any configuration. I have fixed it and set default values for arGlyphs to return original char without calling addGlyphs.
And I have added the 3 patterns to the unit test. 
Please refer to #30.

And I found testEnglishToArabicTransliteration15Cases fails.

**Comment by khaled-alshamaa** [comment]
Thanks for your efforts, I will check it on the weekend and back to you.

**Comment by khaled-alshamaa** [comment]
Dear @sakairyota,

The 3rd pattern doesn't work because you didn't add the letter using `addGlyphs()` method before calling `utf8Glyphs()` for that string:
```php
$Arabic->addGlyphs('ٱ', 'FB50FB51FB50FB51', false, true);
```
Well, because we are in a rush to release version 6.3 of the library, I will merge your PR as an appreciation of your efforts. But, I will remove the code block you have added to the `arGlyphsInit()` because it is not a suitable mechanism for handling the glyphs of Arabic letters context (check this [Arabic Presentation Forms-A](https://en.wikipedia.org/wiki/Arabic_Presentation_Forms-A)). Also, the library code can take properly the characters not listed in the ar_glyphs.json file (but will not process their glyphs if you did not add them using `addGlyphs()`). Still, the new test cases you added to the PHPUnit file are valid, but I need to add the call to the `addGlyphs()` as mentioned before.

**Comment by khaled-alshamaa** [comment]
I fixed the issues in your PR #30

**Comment by sakairyota** [comment]
Dear @khaled-alshamaa

Thanks
